### ISSUE TITLE
Fails to find headers on macOS 10.13

### ISSUE THREAD

**Issue by regexident** [RustAudio/coreaudio-sys#5]
Trying to build **coreaudio-sys** on **macOS 10.13** I get this:

```
$ cargo build

error: failed to run custom build command for `coreaudio-sys v0.2.0 (file:///Users/<snip>/Downloads/coreaudio-sys)`
process didn't exit successfully: `/Users/<snip>/Downloads/coreaudio-sys/target/debug/build/coreaudio-sys-da31a326d38f965f/build-script-build` (exit code: 101)
--- stdout
cargo:rustc-link-lib=framework=AudioToolbox
cargo:rustc-link-lib=framework=AudioUnit
cargo:rustc-link-lib=framework=CoreAudio
cargo:rustc-link-lib=framework=OpenAL
cargo:rustc-link-lib=framework=CoreMIDI

--- stderr
error: header '/System/Library/Frameworks/CoreMIDI.framework/Headers/CoreMIDI.h' does not exist.
thread 'main' panicked at 'unable to generate bindings: ()', src/libcore/result.rs:916:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Peeking into any of the frameworks located in `/System/Library/Frameworks/…` one finds that that there are in fact no headers to be found anywhere inside.

Having installed Xcode however one finds them buried deep down inside of `Xcode.app`:

`/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/…` (`MacOSX` being one of `AppleTVOS`, `AppleTVSimulator`, `iPhoneOS`, `iPhoneSimulator`, `MacOSX`, `WatchOS`, `WatchSimulator`, …)

**Comment by mitchmindtree** [comment]
Hmmm interesting, two macOS 10.13 users have reported successful compilation, however you are [not the first to find the CoreMIDI headers missing from the framework](https://github.com/tomaka/cpal/pull/196#issuecomment-362654033). If you remove the `core_midi` feature, does compilation succeed for you? E.g.
```
cargo build --no-default-features --features "audio_toolbox audio_unit core_audio open_al"
```
does it link and build correctly with this command?

Perhaps we should update the build.rs script so that it checks `Xcode.app` if it can't find the headers in the framework?

**Comment by mitchmindtree** [comment]
cc @HybridEidolon ^

**Comment by regexident** [comment]
Running this:

```bash
cargo build --no-default-features --features "audio_toolbox audio_unit core_audio open_al"
```

Results in:

```
error: header '/System/Library/Frameworks/OpenAL.framework/Headers/alc.h' does not exist.
```

**Comment by mitchmindtree** [comment]
Hmm ok what if you also remove 'open_al' from the list of features?

**Comment by regexident** [comment]
Same.

```
error: header '/System/Library/Frameworks/CoreAudio.framework/Headers/CoreAudio.h' does not exist.
```

None of the frameworks in `/System/Library/Frameworks/` contain any headers for me.
I just checked with a second, individual machine: no headers in said location either.

**Comment by regexident** [comment]
Calling `xcode-select -p` will give you something akin to `/Applications/Xcode.app/Contents/Developer`, from which one could further form the appropriate header file paths.

I reckon something like this might work:

```rust
let platform = if cfg!(target_os = "macos") {
    "MacOSX"
} else if cfg!(target_os = "ios") {
    "iPhoneOS"
} else {
    panic!("Ooops.");
};
let prefix = …; // path from `xcode-select -p`
let infix = format!("Platforms/{}.platform/Developer/SDKs/{}.sdk", platform, platform);
let suffix = "System/Library/Frameworks";
let directory = format!("{}/{}/{}/", prefix, infix, suffix);
```

**Comment by mitchmindtree** [comment]
Just for reference, [this is the Core Audio doc](https://developer.apple.com/library/content/documentation/MusicAudio/Conceptual/CoreAudioOverview/CoreAudioFrameworks/CoreAudioFrameworks.html#//apple_ref/doc/uid/TP40003577-CH9-SW1) that I've been referring to for finding the headers. I wonder if the approach of storing them within the Xcode App itself is the new way?

**Comment by mitchmindtree** [comment]
Thanks for the tip btw! I'd welcome a PR which added a follow-up check for these locations if the file could not be found in the current attempted location using something like you have here! I don't have a mac to test with atm is all, and the one I do have at home already works using the current header paths so wouldn't be useful for testing this anyways.

**Comment by mitchmindtree** [comment]
[This method](https://doc.rust-lang.org/std/path/struct.Path.html#method.exists) is probably the easiest way to test if a header exists, just in case you weren't aware!

**Comment by trsoluti** [comment]
Please merge regexident's fix into your code, as Apple has just recently removed the core audio header files from /System/Library/Frameworks. Thanks.

**Comment by Rhuagh** [comment]
This has been released as version 0.2.1.
